File: app/src/main/java/com/komica/reader/ThreadListActivity.java
package com.komica.reader;

import android.app.AlertDialog;
import android.content.Intent;
import android.os.Bundle;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.View;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageButton;
import android.widget.ProgressBar;
import androidx.annotation.NonNull;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import com.komica.reader.adapter.ThreadAdapter;
import com.komica.reader.model.Board;
import com.komica.reader.model.Thread;
import com.komica.reader.service.KomicaService;
import com.komica.reader.util.FavoritesManager;
import java.util.ArrayList;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.concurrent.ExecutorService;
import java.util.concurrent.Executors;

import androidx.appcompat.widget.Toolbar;
import android.widget.TextView;

public class ThreadListActivity extends AppCompatActivity {

    private RecyclerView recyclerView;
    private ProgressBar progressBar;
    private ThreadAdapter adapter;
    private List<Thread> threads = new ArrayList<>();
    private List<Thread> allThreads = new ArrayList<>();
    private Set<String> existingThreadUrls = new HashSet<>();
    private ExecutorService executor = Executors.newSingleThreadExecutor();
    private Board currentBoard;
    private int currentPage = 0;
    private boolean isLoading = false;
    private boolean hasMore = true;
    private EditText searchEditText;
    private Button sortButton;
    private Button filterButton;
    private ImageButton favoriteButton;
    private String currentSearchQuery = "";
    private FavoritesManager favoritesManager;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_thread_list);

        favoritesManager = FavoritesManager.getInstance(this);

        currentBoard = (Board) getIntent().getSerializableExtra("board");
        
        Toolbar toolbar = findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);
        if (getSupportActionBar() != null) {
            getSupportActionBar().setDisplayShowTitleEnabled(false);
            getSupportActionBar().setDisplayHomeAsUpEnabled(true);
        }
        
        TextView toolbarTitle = findViewById(R.id.toolbarTitle);
        if (currentBoard != null) {
            toolbarTitle.setText(currentBoard.getName());
        }

        recyclerView = findViewById(R.id.recyclerView);
        progressBar = findViewById(R.id.progressBar);
        searchEditText = findViewById(R.id.searchEditText);
        sortButton = findViewById(R.id.sortButton);
        filterButton = findViewById(R.id.filterButton);
        favoriteButton = findViewById(R.id.favoriteButton);

        recyclerView.setLayoutManager(new LinearLayoutManager(this));
        adapter = new ThreadAdapter(threads, new ThreadAdapter.OnThreadClickListener() {
            @Override
            public void onThreadClick(Thread thread) {
                Intent intent = new Intent(ThreadListActivity.this, ThreadDetailActivity.class);
                intent.putExtra("thread", thread);
                startActivity(intent);
            }

            @Override
            public void onShareClick(Thread thread) {
                shareThread(thread);
            }
        });
        recyclerView.setAdapter(adapter);

        setupScrollListener();
        setupSearchListener();
        setupSortListener();
        setupFilterButton();
        setupFavoriteListener();

        if (currentBoard != null) {
            updateFavoriteIcon();
            loadThreads(0);
        }
    }

    private void shareThread(Thread thread) {
        Intent shareIntent = new Intent(Intent.ACTION_SEND);
        shareIntent.setType("text/plain");
        
        String shareText = thread.getTitle() + "\n" + thread.getUrl();
        shareIntent.putExtra(Intent.EXTRA_TEXT, shareText);
        
        startActivity(Intent.createChooser(shareIntent, "分享討論串"));
    }

    private void setupFavoriteListener() {
        favoriteButton.setOnClickListener(v -> {
            if (currentBoard != null) {
                favoritesManager.toggleFavorite(currentBoard.getUrl());
                updateFavoriteIcon();
            }
        });
    }

    private void updateFavoriteIcon() {
        if (currentBoard != null) {
            boolean isFavorite = favoritesManager.isFavorite(currentBoard.getUrl());
            if (isFavorite) {
                favoriteButton.setImageResource(R.drawable.ic_star);
            } else {
                favoriteButton.setImageResource(R.drawable.ic_star_border);
            }
        }
    }

    private void setupFilterButton() {
        filterButton.setOnClickListener(v -> {
            filterThreads();
            android.view.inputmethod.InputMethodManager imm = (android.view.inputmethod.InputMethodManager) getSystemService(android.content.Context.INPUT_METHOD_SERVICE);
            imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
        });
    }

    private void setupScrollListener() {
        LinearLayoutManager layoutManager = (LinearLayoutManager) recyclerView.getLayoutManager();
        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener() {
            @Override
            public void onScrolled(@NonNull RecyclerView recyclerView, int dx, int dy) {
                if (dy > 0) {
                    int visibleItemCount = layoutManager.getChildCount();
                    int totalItemCount = layoutManager.getItemCount();
                    int firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition();

                    if (!isLoading && hasMore && currentSearchQuery.isEmpty()) { // Only load more if not filtering
                        if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount - 5) {
                            loadMoreThreads();
                        }
                    }
                }
            }
        });
    }

    private void setupSearchListener() {
        searchEditText.setOnEditorActionListener((v, actionId, event) -> {
            if (actionId == android.view.inputmethod.EditorInfo.IME_ACTION_DONE) {
                filterThreads();
                android.view.inputmethod.InputMethodManager imm = (android.view.inputmethod.InputMethodManager) getSystemService(android.content.Context.INPUT_METHOD_SERVICE);
                imm.hideSoftInputFromWindow(v.getWindowToken(), 0);
                return true;
            }
            return false;
        });

        searchEditText.addTextChangedListener(new TextWatcher() {
            @Override
            public void beforeTextChanged(CharSequence s, int start, int count, int after) {}

            @Override
            public void onTextChanged(CharSequence s, int start, int before, int count) {}

            @Override
            public void afterTextChanged(Editable s) {
                currentSearchQuery = s.toString().trim().toLowerCase();
                filterThreads();
            }
        });
    }

    private void setupSortListener() {
        sortButton.setOnClickListener(v -> showSortDialog());
    }

    private void showSortDialog() {
        String[] sortOptions = {"最新發表", "最新回覆時間"};
        new AlertDialog.Builder(this)
                .setTitle("選擇排序方式")
                .setItems(sortOptions, (dialog, which) -> {
                    switch (which) {
                        case 0:
                            sortThreadsByLatest();
                            break;
                        case 1:
                            sortThreadsByLastReply();
                            break;
                    }
                    filterThreads();
                })
                .show();
    }

    private void sortThreadsByLastReply() {
        Collections.sort(allThreads, (t1, t2) -> {
            String time1 = t1.getLastReplyTime();
            String time2 = t2.getLastReplyTime();
            if (time1 == null || time1.isEmpty()) return 1;
            if (time2 == null || time2.isEmpty()) return -1;
            return time2.compareTo(time1);
        });
    }

    private void sortThreadsByLatest() {
        Collections.sort(allThreads, Comparator.comparing(Thread::getPostNumber).reversed());
    }

    private void filterThreads() {
        threads.clear();
        if (currentSearchQuery.isEmpty()) {
            threads.addAll(allThreads);
        } else {
            for (Thread thread : allThreads) {
                if (thread.getTitle().toLowerCase().contains(currentSearchQuery) || 
                    thread.getContentPreview().toLowerCase().contains(currentSearchQuery)) {
                    threads.add(thread);
                }
            }
        }
        adapter.notifyDataSetChanged();
    }

    private void loadThreads(int page) {
        if (isLoading) return;

        isLoading = true;
        progressBar.setVisibility(View.VISIBLE);
        executor.execute(() -> {
            try {
                KomicaService.FetchThreadsTask task = new KomicaService.FetchThreadsTask(currentBoard.getUrl(), page);
                List<Thread> newThreads = task.call();

                runOnUiThread(() -> {
                    if (newThreads == null || newThreads.isEmpty()) {
                        hasMore = false;
                        progressBar.setVisibility(View.GONE);
                        isLoading = false;
                    } else {
                        List<Thread> uniqueThreads = new ArrayList<>();
                        for (Thread thread : newThreads) {
                            if (!existingThreadUrls.contains(thread.getUrl())) {
                                existingThreadUrls.add(thread.getUrl());
                                uniqueThreads.add(thread);
                            }
                        }

                        if (uniqueThreads.isEmpty()) {
                            currentPage = page;
                            progressBar.setVisibility(View.GONE);
                            isLoading = false;
                            // Only retry loading next page if we haven't failed too many times consecutively
                            // For simplicity, just stop or try one more time
                            if (page < 10) { // Limit recursion
                                recyclerView.postDelayed(() -> loadMoreThreads(), 500);
                            }
                        } else {
                            currentPage = page;
                            allThreads.addAll(uniqueThreads);
                            sortThreadsByLatest();
                            filterThreads();
                            progressBar.setVisibility(View.GONE);
                            isLoading = false;
                        }
                    }
                });
            } catch (Exception e) {
                e.printStackTrace();
                runOnUiThread(() -> {
                    progressBar.setVisibility(View.GONE);
                    isLoading = false;
                });
            }
        });
    }

    private void loadMoreThreads() {
        currentPage++;
        loadThreads(currentPage);
    }

    @Override
    protected void onDestroy() {
        super.onDestroy();
        executor.shutdown();
    }
}

File: app/src/main/java/com/komica/reader/ThreadListActivity.kt
package com.komica.reader

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.text.Editable
import android.text.TextWatcher
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.view.inputmethod.EditorInfo
import android.view.inputmethod.InputMethodManager
import androidx.activity.viewModels
import androidx.appcompat.app.AppCompatActivity
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.komica.reader.adapter.ThreadAdapter
import com.komica.reader.databinding.ActivityThreadListBinding
import com.komica.reader.model.Board
import com.komica.reader.model.Thread
import com.komica.reader.util.FavoritesManager
import com.komica.reader.util.getSerializableCompat
import com.komica.reader.viewmodel.ThreadListViewModel
import dagger.hilt.android.AndroidEntryPoint
import javax.inject.Inject

@AndroidEntryPoint
class ThreadListActivity : AppCompatActivity() {

    private lateinit var binding: ActivityThreadListBinding
    private lateinit var adapter: ThreadAdapter
    private val threads: MutableList<Thread> = ArrayList()
    private val viewModel: ThreadListViewModel by viewModels()
    private var currentBoard: Board? = null
    
    @Inject
    lateinit var favoritesManager: FavoritesManager

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = ActivityThreadListBinding.inflate(layoutInflater)
        setContentView(binding.root)

        // favoritesManager injected by Hilt
        
        currentBoard = intent.getSerializableCompat("board")

        if (currentBoard == null) {
            finish()
            return
        }

        setSupportActionBar(binding.toolbar)
        supportActionBar?.apply {
            setDisplayShowTitleEnabled(false)
            setDisplayHomeAsUpEnabled(true)
        }

        binding.toolbarTitle.text = currentBoard!!.name

        binding.recyclerView.layoutManager = LinearLayoutManager(this)

        binding.swipeRefreshLayout.setOnRefreshListener {
            viewModel.refresh()
        }

        adapter = ThreadAdapter(threads, object : ThreadAdapter.OnThreadClickListener {
            override fun onThreadClick(thread: Thread) {
                val intent = Intent(this@ThreadListActivity, ThreadDetailActivity::class.java)
                intent.putExtra("thread_url", thread.url)
                intent.putExtra("thread_title", thread.title)
                startActivity(intent)
            }

            override fun onShareClick(thread: Thread) {
                shareThread(thread)
            }
        })
        binding.recyclerView.adapter = adapter

        setupScrollListener()
        setupSearchListener()
        setupFilterButton()
        setupFavoriteListener()

        updateFavoriteIcon()
        observeViewModel()
    }

    private fun observeViewModel() {
        viewModel.threads.observe(this) { newThreads ->
            adapter.updateThreads(newThreads)
            binding.swipeRefreshLayout.isRefreshing = false
        }

        viewModel.isLoading.observe(this) { isLoading ->
            if (!binding.swipeRefreshLayout.isRefreshing) {
                val isEmpty = threads.isEmpty()
                binding.progressBar.visibility = if (isLoading && isEmpty) View.VISIBLE else View.GONE
            }
        }

        viewModel.errorMessage.observe(this) { error ->
            if (error != null) {
                showErrorSnackBar(error)
            }
        }
    }

    private fun showErrorSnackBar(message: String) {
        Snackbar.make(binding.recyclerView, message, Snackbar.LENGTH_LONG)
            .setAction(R.string.action_retry) { viewModel.refresh() }
            .show()
    }

    override fun onResume() {
        super.onResume()
        if (::adapter.isInitialized) {
            adapter.notifyDataSetChanged()
        }
    }

    private fun shareThread(thread: Thread) {
        val shareIntent = Intent(Intent.ACTION_SEND).apply {
            type = "text/plain"
            val shareText = "${thread.title}\n${thread.url}"
            putExtra(Intent.EXTRA_TEXT, shareText)
        }
        startActivity(Intent.createChooser(shareIntent, getString(R.string.action_share_thread)))
    }

    private fun setupFavoriteListener() {
        binding.favoriteButton.setOnClickListener {
            currentBoard?.url?.let { url ->
                favoritesManager.toggleFavorite(url)
                updateFavoriteIcon()
            }
        }
    }

    private fun updateFavoriteIcon() {
        currentBoard?.url?.let { url ->
            val isFavorite = favoritesManager.isFavorite(url)
            binding.favoriteButton.setImageResource(if (isFavorite) R.drawable.ic_star else R.drawable.ic_star_border)
        }
    }

    private fun setupFilterButton() {
        binding.filterButton.setOnClickListener { v ->
            val query = binding.searchEditText.text.toString().trim()
            if (query.isNotEmpty()) {
                viewModel.setSearchQuery(query)
                viewModel.performRemoteSearch()
            } else {
                viewModel.clearSearch()
            }
            hideKeyboard(v)
        }
    }

    private fun hideKeyboard(v: View) {
        val imm = getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
        imm?.hideSoftInputFromWindow(v.windowToken, 0)
    }

    private fun setupScrollListener() {
        val layoutManager = binding.recyclerView.layoutManager as LinearLayoutManager
        binding.recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 0) {
                    val visibleItemCount = layoutManager.childCount
                    val totalItemCount = layoutManager.itemCount
                    val firstVisibleItemPosition = layoutManager.findFirstVisibleItemPosition()

                    if ((visibleItemCount + firstVisibleItemPosition) >= totalItemCount - 5) {
                        viewModel.loadMore()
                    }
                }
            }
        })
    }

    private fun setupSearchListener() {
        binding.searchEditText.setOnEditorActionListener { v, actionId, _ ->
            if (actionId == EditorInfo.IME_ACTION_DONE || actionId == EditorInfo.IME_ACTION_SEARCH) {
                val query = binding.searchEditText.text.toString().trim()
                if (query.isNotEmpty()) {
                    viewModel.setSearchQuery(query)
                    viewModel.performRemoteSearch()
                } else {
                    viewModel.clearSearch()
                }
                hideKeyboard(v)
                true
            } else {
                false
            }
        }

        binding.searchEditText.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(s: CharSequence?, start: Int, count: Int, after: Int) {}
            override fun onTextChanged(s: CharSequence?, start: Int, before: Int, count: Int) {}
            override fun afterTextChanged(s: Editable?) {
                val query = s.toString().trim()
                viewModel.setSearchQuery(query)
                if (query.isEmpty()) {
                    viewModel.clearSearch()
                }
            }
        })
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.thread_list_menu, menu)
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem): Boolean {
        return when (item.itemId) {
            android.R.id.home -> {
                finish()
                true
            }
            R.id.action_history -> {
                val intent = Intent(this, HistoryActivity::class.java)
                startActivity(intent)
                true
            }
            else -> super.onOptionsItemSelected(item)
        }
    }
}
